File: Shared/ConduitKit/Sources/ConduitProtocol/Frame.swift
//
//  Frame.swift
//  ConduitProtocol
//
//  Conduit Link framing. Spec: Shared/Protocol/README.md §2.
//
//      ┌─────────┬─────────┬───────────────────┬─────────────────┐
//      │ channel │  type   │   length (u32 BE) │ payload         │
//      └─────────┴─────────┴───────────────────┴─────────────────┘
//
//  The header and the channel layouts come from the trackpad prototype that
//  preceded Conduit, measured on real hardware. There is no magic number and
//  no resynchronisation: TCP neither drops nor reorders bytes, so an unknown
//  channel or type can only mean a bug, and guessing past it would turn
//  garbage into input events.
//

import Foundation

public enum LinkChannel: UInt8, Sendable, CaseIterable {
    case control = 0
    case input = 1
    case sensor = 2
    case haptic = 3
    case media = 4
    case file = 5

    /// Largest payload accepted on this channel.
    public var maxPayloadSize: Int {
        self == .control ? 1 << 20 : 4 << 20
    }

    /// Frame types defined on this channel in protocol v1. Reserved channels
    /// define none yet, so any frame on them is rejected.
    public var knownTypes: Set<UInt8> {
        switch self {
        case .control: return [ControlType.handshake.rawValue, ControlType.envelope.rawValue]
        case .input:   return Set(InputType.allCases.map(\.rawValue))
        case .haptic:  return [HapticType.vibrate.rawValue]
        case .sensor, .media, .file: return []
        }
    }
}

public enum ControlType: UInt8, Sendable {
    /// Plaintext; only before the session is established.
    case handshake = 0
    /// A sealed Envelope; only after.
    case envelope = 1
}

public struct Frame: Sendable, Equatable {
    public static let headerSize = 6

    public let channel: LinkChannel
    public let type: UInt8
    public let payload: Data

    public init(channel: LinkChannel, type: UInt8, payload: Data = Data()) {
        self.channel = channel
        self.type = type
        self.payload = payload
    }

    public func encoded() -> Data {
        var out = Data(capacity: Self.headerSize + payload.count)
        out.append(channel.rawValue)
        out.append(type)
        let length = UInt32(payload.count)
        out.append(UInt8(truncatingIfNeeded: length >> 24))
        out.append(UInt8(truncatingIfNeeded: length >> 16))
        out.append(UInt8(truncatingIfNeeded: length >> 8))
        out.append(UInt8(truncatingIfNeeded: length))
        out.append(payload)
        return out
    }
}

public enum FrameError: Error, Sendable, Equatable {
    case unknownChannel(UInt8)
    case unknownType(channel: LinkChannel, type: UInt8)
    case tooLarge(channel: LinkChannel, length: Int)
}

/// Incremental frame decoder. TCP delivers arbitrary chunks, so bytes are
/// buffered and every complete frame is returned; the remainder waits for
/// the next call. After an error the decoder refuses further input — the
/// stream position can no longer be trusted.
public struct FrameDecoder: Sendable {

    private var buffer = Data()
    public private(set) var failure: FrameError?

    public init() {}

    public mutating func receive(_ data: Data) throws -> [Frame] {
        if let failure { throw failure }
        buffer.append(data)

        var frames: [Frame] = []
        while buffer.count >= Frame.headerSize {
            let base = buffer.startIndex
            let rawChannel = buffer[base]
            let type = buffer[base + 1]
            let length = Int(buffer[base + 2]) << 24 | Int(buffer[base + 3]) << 16
                       | Int(buffer[base + 4]) << 8 | Int(buffer[base + 5])

            guard let channel = LinkChannel(rawValue: rawChannel) else {
                return try fail(.unknownChannel(rawChannel))
            }
            guard channel.knownTypes.contains(type) else {
                return try fail(.unknownType(channel: channel, type: type))
            }
            guard length <= channel.maxPayloadSize else {
                return try fail(.tooLarge(channel: channel, length: length))
            }
            guard buffer.count >= Frame.headerSize + length else { break }

            let start = base + Frame.headerSize
            frames.append(Frame(channel: channel, type: type, payload: Data(buffer[start ..< start + length])))
            buffer.removeSubrange(base ..< start + length)
        }
        return frames
    }

    private mutating func fail(_ error: FrameError) throws -> [Frame] {
        failure = error
        buffer.removeAll()
        throw error
    }
}

File: Android/core/src/main/java/com/khushi/conduit/core/protocol/Frame.kt
package com.khushi.conduit.core.protocol

import java.io.ByteArrayOutputStream

/**
 * Conduit Link framing. Spec: Shared/Protocol/README.md §2.
 *
 *     [channel u8][type u8][length u32 BE][payload]
 *
 * No magic number and no resynchronisation: TCP neither drops nor reorders
 * bytes, so an unknown channel or type can only mean a bug, and guessing past
 * it would turn garbage into input events.
 */
enum class LinkChannel(val value: Int) {
    CONTROL(0),
    INPUT(1),
    SENSOR(2),
    HAPTIC(3),
    MEDIA(4),
    FILE(5);

    val maxPayloadSize: Int get() = if (this == CONTROL) 1 shl 20 else 4 shl 20

    /** Frame types defined in protocol v1. Reserved channels define none yet. */
    val knownTypes: Set<Int>
        get() = when (this) {
            CONTROL -> setOf(ControlType.HANDSHAKE, ControlType.ENVELOPE)
            INPUT -> InputType.entries.map { it.value }.toSet()
            HAPTIC -> setOf(HapticType.VIBRATE)
            SENSOR, MEDIA, FILE -> emptySet()
        }

    companion object {
        fun of(value: Int): LinkChannel? = entries.firstOrNull { it.value == value }
    }
}

object ControlType {
    /** Plaintext; only before the session is established. */
    const val HANDSHAKE = 0
    /** A sealed Envelope; only after. */
    const val ENVELOPE = 1
}

class Frame(val channel: LinkChannel, val type: Int, val payload: ByteArray = ByteArray(0)) {

    fun encode(): ByteArray {
        val out = ByteArrayOutputStream(HEADER_SIZE + payload.size)
        out.write(channel.value)
        out.write(type)
        val length = payload.size
        out.write(length ushr 24 and 0xFF)
        out.write(length ushr 16 and 0xFF)
        out.write(length ushr 8 and 0xFF)
        out.write(length and 0xFF)
        out.write(payload)
        return out.toByteArray()
    }

    override fun equals(other: Any?): Boolean =
        other is Frame && other.channel == channel && other.type == type && other.payload.contentEquals(payload)

    override fun hashCode(): Int = (channel.hashCode() * 31 + type) * 31 + payload.contentHashCode()

    override fun toString(): String = "Frame($channel, type=$type, ${payload.size} bytes)"

    companion object {
        const val HEADER_SIZE = 6
    }
}

class FrameException(val reason: Reason, message: String) : Exception(message) {
    enum class Reason { UNKNOWN_CHANNEL, UNKNOWN_TYPE, TOO_LARGE }
}

/**
 * Incremental frame decoder. TCP delivers arbitrary chunks, so bytes are
 * buffered and every complete frame is returned. After an error the decoder
 * refuses further input — the stream position can no longer be trusted.
 */
class FrameDecoder {
    private var buffer = ByteArray(0)
    var failure: FrameException? = null
        private set

    fun receive(data: ByteArray): List<Frame> {
        failure?.let { throw it }
        buffer += data

        val frames = mutableListOf<Frame>()
        var offset = 0
        while (buffer.size - offset >= Frame.HEADER_SIZE) {
            val rawChannel = buffer[offset].toInt() and 0xFF
            val type = buffer[offset + 1].toInt() and 0xFF
            val length = (buffer[offset + 2].toInt() and 0xFF shl 24) or
                (buffer[offset + 3].toInt() and 0xFF shl 16) or
                (buffer[offset + 4].toInt() and 0xFF shl 8) or
                (buffer[offset + 5].toInt() and 0xFF)

            val channel = LinkChannel.of(rawChannel)
                ?: fail(FrameException.Reason.UNKNOWN_CHANNEL, "unknown channel $rawChannel")
            if (type !in channel.knownTypes) fail(FrameException.Reason.UNKNOWN_TYPE, "unknown type $type on $channel")
            // A negative length means the u32 exceeded Int.MAX_VALUE: too large either way.
            if (length < 0 || length > channel.maxPayloadSize) {
                fail(FrameException.Reason.TOO_LARGE, "payload of $length bytes on $channel")
            }
            if (buffer.size - offset < Frame.HEADER_SIZE + length) break

            val start = offset + Frame.HEADER_SIZE
            frames += Frame(channel, type, buffer.copyOfRange(start, start + length))
            offset = start + length
        }
        buffer = buffer.copyOfRange(offset, buffer.size)
        return frames
    }

    private fun fail(reason: FrameException.Reason, message: String): Nothing {
        val error = FrameException(reason, message)
        failure = error
        buffer = ByteArray(0)
        throw error
    }
}
